### PR TITLE
introduce policyfilter mechanism

### PR DESCRIPTION
**Pull request by kkourt**
This PR introduces the policyfilter mechanism, both from bpf- and user-space side. It is currently unused, but it will be used in subsequent PRs for implementing namespaced tracing policies.

* Depends on: #700
* Part of: #694

### CODE REVIEW

File: pkg/policyfilter/state.go [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+//  - periodic (or via external command) rescans to ensure that everything is up-to-date
+//  - optimization: policy<->pod matching caching so that we can go from pod -> matching policies
+//  without having to check all policies. Note that this would have to account for new
+//  policies added. Since we are only matching on namespace, we iterate over all policies for now
+//  since the performance impact of above optimization would be small.
+
+const (
+	// polMapSize is the number of entries for the (inner) policy map. It
+	// should be large enough to accommodate the number of containers
+	// running in a systme.
```
Comment (tpapagian) [inline pkg/policyfilter/state.go:57]: typo: systme -> system

File: pkg/policyfilter/map.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+	ret, err := ebpf.NewMap(policyMapSpec)
+	if err != nil {
+		return PfMap{}, err
+	}
+
+	mapDir := bpf.MapPrefixPath()
+	pinPath := filepath.Join(mapDir, MapName)
+	os.Remove(pinPath)
+	os.Mkdir(mapDir, os.ModeDir)
```
Comment (tpapagian) [inline pkg/policyfilter/map.go:57]: Here you just remove the map. Why do you need to call `mkdir` again?

File: pkg/cgroups/cgroups.go [hunk 3]
```diff
@@ -699,3 +700,64 @@ func CgroupNameFromCStr(cstr []byte) string {
 	}
 	return string(cstr[:i])
 }
+
+func tryHostCgoup(path string) error {
```
Comment (tpapagian) [inline pkg/cgroups/cgroups.go:704]: typo: tryHostCgoup -> tryHostCgroup

**Review by tpapagian — COMMENTED** [review]
Thanks! I have added some initial comments from my side.

File: bpf/process/policy_filter.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	curr = execve_map_get_noinit(pid);
+	if (!curr || curr->cgrpid_tracker == 0)
+		return false;
+
+	policy_map = map_lookup_elem(&policy_filter_maps, &policy_id);
+	if (!policy_map)
+		return false;
+
+	policy_val = map_lookup_elem(policy_map, &curr->cgrpid_tracker);
+	return policy_val;
```
[the remark this answers is not in the document]
Comment (kkourt) [reply bpf/process/policy_filter.h:47]: makes sense -- changed it.

File: pkg/policyfilter/map.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+// PfMap is a simple wrapper for ebpf.Map so that we can write methods for it
+type PfMap struct {
+	*ebpf.Map
+}
+
+// newMap returns a new policy filter map.
+func newPfMap() (PfMap, error) {
+	// use the generic kprobe program, to find the policy filter map spec
+	objName, _ := kernels.GenericKprobeObjs()
```
[the remark this answers is not in the document]
Comment (kkourt) [reply pkg/policyfilter/map.go:33]: I don't see why it wouldn't. We just need the spec for the map, which should be the same on all the objects.

File: pkg/policyfilter/state.go (hunk 1 above)
Comment (kkourt) [reply pkg/policyfilter/state.go:57]: fixed, thanks!

File: pkg/policyfilter/map.go (hunk 2 above)
Comment (kkourt) [reply pkg/policyfilter/map.go:57]: There are cases where the map does not exist (os.Remove() will just return an error, which we do not check) and neither does the directory. We could check if the directory exists or not but not sure if it's worth it.

File: pkg/cgroups/cgroups.go (hunk 3 above)
Comment (kkourt) [reply pkg/cgroups/cgroups.go:704]: thanks, fixed.

File: pkg/policyfilter/map.go (hunk 2 above)
Comment (tpapagian) [reply pkg/policyfilter/map.go:57]: Sounds reasonable, thanks!

File: pkg/policyfilter/map.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tpapagian) [reply pkg/policyfilter/map.go:33]: Ok, I see. I miss-read that part.

File: bpf/process/bpf_generic_kprobe.c [hunk 6]
```diff
@@ -90,6 +91,10 @@ generic_kprobe_start_process_filter(void *ctx)
 	config = map_lookup_elem(&config_map, &msg->idx);
 	if (!config)
 		return 0;
+
+	if (!policy_filter_check(config->policy_id))
+		return 0;
```
[the remark this answers is not in the document]
Comment (kkourt) [reply bpf/process/bpf_generic_kprobe.c]: Makes sense. Fixed it!

**Review by tpapagian — APPROVED** [review]
LGTM, thanks!

**Review by willfindlay — APPROVED** [review]

**Pull request merged**